### ISSUE TITLE
how do you get the string out of  Tensor("args_0:0", shape=(), dtype=string) type?:

### ISSUE THREAD

**Issue by SlowMonk** [tensorflow/tensorflow#40309]
I tried to get the string out of 
type  tensor("args_0:0", shape=(), dtype=string). 

The reason I tried to do is that I want to get the original image size 

```

 dataset = tf.data.Dataset.from_tensor_slices((images,boxes,labels))
    run_train(dataset.map(resize_image_bbox, num_parallel_calls=tf.data.experimental.AUTOTUNE).batch(1).prefetch(tf.data.experimental.AUTOTUNE))

```

this images is the list of image_path 
then I passed it to the 

```
def resize_image_bbox(image,boxes,labels):

    print('image->',image)
    img = tf.io.read_file(image)
    img = tf.image.decode_jpeg(img, channels=3)
    img = tf.cast(img, tf.float32)
    newSize = (300, 300)

    new_img = tf.image.resize(img,newSize)
    return new_img,boxes,labels
```
I also need to resize my boxes with original image size
but I have no idea to get the original image size from it 

so I tried to read it from  the path then get the original image size 

If you have better idea let me know
=============================================================
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): ubuntu 19.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source 
- TensorFlow version (use command below): 2.0 
- Python version: 2.0 
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 10.1
- GPU model and memory: titan xp 

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

**Describe the expected behavior**

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by amahendrakar** [comment]
@SlowMonk,
You can use the `.numpy()` function to extract the string from the Tensor. Please check [this gist](https://colab.research.google.com/gist/amahendrakar/4645fa442bcbcf63365b8390ce75c575/40309.ipynb) for reference. 

Also, if this issue is not a TensorFlow bug or feature request, it is better asked on [StackOverflow](https://stackoverflow.com/questions/tagged/tensorflow). Since there is a larger community that reads questions there. Thanks!

**Comment by pn12** [comment]
@SlowMonk  - Did you get a way to extract value?

**Comment by shelllet** [comment]
The same question

**Comment by 2er0** [comment]
Workaround: https://stackoverflow.com/questions/58969880/how-to-use-map-with-tuples-in-a-tensorflow-2-dataset

**Comment by microcoder-py** [comment]
Won't work because when we use the map function it passes a Tensor object and not EagerTensor which has no .numpy() function

Please provide recourse, thanks

**Comment by marcocaccin** [comment]
This issue is in no way resolved and should not be closed.

**Comment by janeusz2000** [comment]
same here... I try to parse file_path to get the label value from the map, which is a float value and because of that I am unable to do so

**Comment by kaomoneus** [comment]
same here. TF Team,  it's quite weird that you provided so many tools to convert python POD into tensors, and no way to get it back.

**Comment by dubrovin-sudo** [comment]
Same problem...

**Comment by TeamDman** [comment]
I just switched to Pytorch tbh, never looked back.

**Comment by arijitmeesho** [comment]
This is very weird ..print statement of tensorflow can't be believed

**Comment by Bourhano** [comment]
a wise wizard once said: 'everything is not what it seems'

**Comment by snps-kyuhyun** [comment]
Hi, all. I faced same issue and finally got a solution from the link below.

https://stackoverflow.com/questions/56122670/how-to-get-string-value-out-of-tf-tensor-which-dtype-is-string

sample code:
```Python
from PIL import Image
import tensorflow as tf

def load_file(filename):
  return img = tf.constant(np.array(Image.open(x.numpy()).convert("L")))

ds = tf.data.Dataset.list_files("/*.png")
ds_map = ds.map(lambda x: tf.py_function(load_file, [x], [tf.uint8]))
```

Try `tf.py_function`!

**Comment by YuxuanXie** [comment]
Same problem...

**Comment by AlienKevin** [comment]
As mentioned by @snps-kyuhyun, the solution is to use `tf.py_function`. I'm gonna elaborate on the process a bit. Essentially, `Tensor("args_0:0")` shows that this is a symbolic tensor meaning it's only a placeholder in the compute graph. The value of the tensor is only determined when you pump data through the graph (i.e. when you load data points from the dataset). In other words, `Tensor("args_0:0")` is a lazy tensor that won't get evaluated until you need its value.

Most of the time, we don't need to know the value stored in the tensor to construct the compute graph. This is because all tensor operations can take either symbolic or value tensors. However, the problem arises when you try to implement logic based on the *content of the tensor*. Like you want to read the file path stored in a symbolic string tensor. TensorFlow recognizes these kind of use cases by providing a generic `tf.py_function` api. It's basically a glorified lambda for the compute graph. If you look at `py_function`'s signature, you will see that it requires in three arguments:
```
func: A Python function that accepts the value of inp as arguments, and returns a Tensor
inp: the input symbolic tensor
Tout: the data type used for the output tensor
```
The key argument `func` is basically any custom operation you want to do on the `inp` symbolic tensor. `py_function` is a way to defer the logic in `func` until you know the value of the `inp`.

I'm gonna include the previous sample code as an example:
```
from PIL import Image
import tensorflow as tf

def load_file(filename):
  return img = tf.constant(np.array(Image.open(x.numpy()).convert("L")))

ds = tf.data.Dataset.list_files("/*.png")
ds_map = ds.map(lambda x: tf.py_function(load_file, [x], [tf.uint8]))
```

Why do we need to use `tf.py_function` here? Let's try the plain version without `py_function` and see how it breaks down:
```
ds = tf.data.Dataset.list_files("/*.png")
ds_map = ds.map(lambda x: load_file(x))
```

Python will throw an error saying that `numpy` cannot be called on `Tensor` object, because `x` is a symbolic tensor. How do you convert a symbolic tensor to a value tensor? By wrapping the `load_file` with `py_function`, you can get the value out of the symbolic tensor `x` once it's evaluated.

**Comment by sun8608** [comment]
The problem with tf.py_function is that it does not allow dict output from the given func, where dict is a common data structure that is used nearly everywhere with other tf functions, you have to flattern the dict and convert it back for follow up processing steps, which is really annoying there

**Comment by sequoiarose** [comment]
While the solutions above by @AlienKevin and @snps-kyuhyun seem to load the data, the dataset object looses its shape from tf.py_function (see https://stackoverflow.com/questions/65321954/tensorflow-custom-preprocessing-with-tf-py-function-losing-shape)

If anyone knows how to add the shape back to the dataset please let me know. The S/O post is not in the dataset object so it is a bit unclear.

This inefficiency has been plaguing me for days. Tensorflow really needs a better solution for loading and processing large custom datasets. All the examples available are for classic datasets and not useful for custom data loading and processing. I'm about to scrap all my code and start transitioning it all to pytorch because this is ridiculous and beyond frustrating.
